### PR TITLE
Add cdp executor

### PR DESCRIPTION
**Pull request by brunotm**
This change adds a cdp "proxy" and a small run api that also manages the lifecycle of the underlying chrome process. And will allow us to move forward in a stable testing environment.

This change is the first iteration for decoupling the server component from the test execution which will be tracked on https://github.com/Unbabel/replicant/issues/6

### CODE REVIEW

File: internal/cdpserver/replicant-cdp.go [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+		select {
+		case <-time.After(s.interval):
+
+			log.Info("recycling chrome process").Int("pid", int64(s.cmd.Process.Pid)).Log()
+			s.m.Lock()
+
+			// stop chrome process and its children
+			err := syscall.Kill(-s.cmd.Process.Pid, syscall.SIGKILL)
+			if err != nil {
+				return fmt.Errorf("replicant-cdp: error stopping chrome process: %w", err)
```
Comment (brunotm) [inline internal/cdpserver/replicant-cdp.go]: Codacy Issue found: [ unrecognized printf verb 'w'](https://app.codacy.com/gh/Unbabel/replicant/pullRequest?prid=4990900)

**Comment by brunotm** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 7
           

Complexity increasing per file
==============================
- driver/web/transaction.go  5
- cmd/replicant-cdp/main.go  4
- internal/cdpserver/replicant-cdp.go  7
         

Clones added
============
- internal/cdpserver/replicant-cdp.go  3
         

Clones removed
==============
+ api/run.go  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/Unbabel/replicant/pullRequest?prid=4990900&bid=16476279)

File: driver/web/transaction.go [hunk 2]
```diff
@@ -71,29 +75,62 @@ func (t *Transaction) Run(ctx context.Context) (result transaction.Result) {
...
+
+		buf, err = ioutil.ReadAll(resp.Body)
+		if err != nil {
+			result.Error = fmt.Errorf("driver/web: error reading proxied result data: %w", err)
+			result.Failed = true
+			break
+		}
+
+		if err = json.Unmarshal(buf, &result); err != nil {
+			result.Error = fmt.Errorf("driver/web: error deserializing proxied result data: %w", err)
```
Comment (brunotm) [inline driver/web/transaction.go]: Codacy Issue found: [ unrecognized printf verb 'w'](https://app.codacy.com/gh/Unbabel/replicant/pullRequest?prid=4990900)

File: internal/cdpserver/replicant-cdp.go [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+		var buf []byte
+		var config transaction.Config
+
+		if buf, err = ioutil.ReadAll(r.Body); err != nil {
+			httpError(w, fmt.Errorf("error reading request body: %w", err), http.StatusBadRequest)
+			return
+		}
+
+		if err = json.Unmarshal(buf, &config); err != nil {
+			httpError(w, fmt.Errorf("error deserializing json request body: %w", err), http.StatusBadRequest)
```
Comment (brunotm) [inline internal/cdpserver/replicant-cdp.go]: Codacy Issue found: [ unrecognized printf verb 'w'](https://app.codacy.com/gh/Unbabel/replicant/pullRequest?prid=4990900)

File: internal/cdpserver/replicant-cdp.go [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+		s.m.RLock()
+		defer s.m.RUnlock()
+		defer r.Body.Close()
+
+		var err error
+		var buf []byte
+		var config transaction.Config
+
+		if buf, err = ioutil.ReadAll(r.Body); err != nil {
+			httpError(w, fmt.Errorf("error reading request body: %w", err), http.StatusBadRequest)
```
Comment (brunotm) [inline internal/cdpserver/replicant-cdp.go]: Codacy Issue found: [ unrecognized printf verb 'w'](https://app.codacy.com/gh/Unbabel/replicant/pullRequest?prid=4990900)

File: internal/cdpserver/replicant-cdp.go [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+			if err != nil {
+				return fmt.Errorf("replicant-cdp: error stopping chrome process: %w", err)
+			}
+
+			// start chrome process and set process group id to avoid
+			// leaving zombies upon termination
+			s.cmd = exec.Command(arguments[0], arguments[1:]...)
+			s.cmd.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}
+			if err := s.cmd.Start(); err != nil {
+				return fmt.Errorf("replicant-cdp: error starting chrome process: %w", err)
```
Comment (brunotm) [inline internal/cdpserver/replicant-cdp.go]: Codacy Issue found: [ unrecognized printf verb 'w'](https://app.codacy.com/gh/Unbabel/replicant/pullRequest?prid=4990900)

File: internal/cdpserver/replicant-cdp.go [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+
+		log.Info("handling proxied transaction request").String("name", config.Name).Log()
+
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+
+		result := tx.Run(ctx)
+		buf, err = json.Marshal(&result)
+		if err != nil {
+			httpError(w, fmt.Errorf("error serializing results: %w", err), http.StatusInternalServerError)
```
Comment (brunotm) [inline internal/cdpserver/replicant-cdp.go]: Codacy Issue found: [ unrecognized printf verb 'w'](https://app.codacy.com/gh/Unbabel/replicant/pullRequest?prid=4990900)

File: driver/web/transaction.go [hunk 7]
```diff
@@ -71,29 +75,62 @@ func (t *Transaction) Run(ctx context.Context) (result transaction.Result) {
...
+		if err != nil {
+			result.Error = err
+			result.Failed = true
+			break
+		}
+		defer resp.Body.Close()
+
+		buf, err = ioutil.ReadAll(resp.Body)
+		if err != nil {
+			result.Error = fmt.Errorf("driver/web: error reading proxied result data: %w", err)
```
Comment (brunotm) [inline driver/web/transaction.go]: Codacy Issue found: [ unrecognized printf verb 'w'](https://app.codacy.com/gh/Unbabel/replicant/pullRequest?prid=4990900)

File: internal/cdpserver/replicant-cdp.go (hunk 1 above)
Comment (brunotm) [reply internal/cdpserver/replicant-cdp.go]: This is a false positive since the 'w' verb is implemented since go 1.13 for error wrapping.

**Comment by brunotm** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
           

Complexity increasing per file
==============================
- cmd/replicant-cdp/main.go  4
- internal/cdpserver/replicant-cdp.go  7
- driver/web/transaction.go  5
         

Clones added
============
- internal/cdpserver/replicant-cdp.go  3
         

Clones removed
==============
+ api/run.go  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/Unbabel/replicant/pullRequest?prid=4990900&bid=16476279)

File: driver/web/transaction.go [hunk 8]
```diff
@@ -71,29 +75,62 @@ func (t *Transaction) Run(ctx context.Context) (result transaction.Result) {
...
+			result.Failed = true
+			break
+		}
+
+		if err = json.Unmarshal(buf, &result); err != nil {
+			result.Error = fmt.Errorf("driver/web: error deserializing proxied result data: %w", err)
```
Comment (brunotm) [inline driver/web/transaction.go:126]: Codacy Issue found: [ unrecognized printf verb 'w'](https://app.codacy.com/gh/Unbabel/replicant/pullRequest?prid=4990900)

**Comment by brunotm** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- driver/web/transaction.go  5
- cmd/replicant-cdp/main.go  4
- internal/cdpserver/replicant-cdp.go  7
         

Clones added
============
- internal/cdpserver/replicant-cdp.go  3
         

Clones removed
==============
+ api/run.go  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/Unbabel/replicant/pullRequest?prid=4990900&bid=16476279)

File: driver/web/transaction.go [hunk 9]
```diff
@@ -71,29 +75,62 @@ func (t *Transaction) Run(ctx context.Context) (result transaction.Result) {
...
+		if err != nil {
+			result.Error = err
+			result.Failed = true
+			break
+		}
+		defer resp.Body.Close()
+
+		buf, err = ioutil.ReadAll(resp.Body)
+		if err != nil {
+			result.Error = fmt.Errorf("driver/web: error reading proxied result data: %w", err)
```
Comment (brunotm) [inline driver/web/transaction.go:120]: Codacy Issue found: [ unrecognized printf verb 'w'](https://app.codacy.com/gh/Unbabel/replicant/pullRequest?prid=4990900)

**Review by gonzalesraul — APPROVED** [review]
LGTM

**Pull request merged**